File: 1319-number-of-operations-to-make-network-connected/1319-number-of-operations-to-make-network-connected.cpp
class Solution {
public:
    unordered_map<int,int>par, rank;
    int findd(int x){
        if(par[x] == x)return x;
        return par[x] = findd(par[x]);
    }
    void unionn(int u,int v){
        u = findd(u);
        v = findd(v);
        if(rank[u] > rank[v]){
            par[v] = u;
            rank[u] += rank[v];
        }
        else{
            par[u] = v;
            rank[v] += rank[u];
        }
    }
    int makeConnected(int n, vector<vector<int>>& connections) {
        for(int i=0;i<n;i++){
            par[i] = i;
            rank[i]++;
        }
        int cnt = 0;
        for(auto it : connections){
            int u = it[0], v = it[1];
            if(findd(u) == findd(v))cnt++;
            else unionn(u,v);
        }
        unordered_set<int>s;
        for(int i=0;i<n;i++){
            s.insert(findd(i));
        }
        int need = s.size() - 1;
        if(need > cnt)return -1;
        return need;
    }
};

File: 1319-number-of-operations-to-make-network-connected/1319-number-of-operations-to-make-network-connected.py
class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        c=[len(connections)]
        vis=[0]*n
        par=[i for i in range(n)]
        def findd(x):
            if x==par[x]:
                return x
            else:
                par[x]=findd(par[x])
                return par[x]
        def union(a,b):
            x=findd(a)
            y=findd(b)
            if x!=y:
                c[0]-=1
            par[y]=x
        for i in connections:
            union(i[0],i[1])
            vis[i[0]]=1
            vis[i[1]]=1
        d=set()
        for i in range(n):
            d.add(findd(i))
        d = len(d) - 1
        if d>c[0]:
            return -1
        
        return d